assignment3/word_search.py

```diff
@@ -0,0 +1,93 @@
+#! /usr/bin/python3.5
+
+import dictionary
+ALL_WORDS = set()
+WORD_DICT = None
+CHAR_2D_ARRAY = None
+
+def findAllWords(row, col, char2dArray, wordDict):
+    """Find all valid words that can be formed in the char2dArray. 
+    Note that CHAR_2D_ARRAY, WORD_DICT and ALL_WORDS are global variables. 
+
+    Args:
+        row: no.rows in char2dArray
+        col: no.col in char2dArray
+        char2dArray: the grid of letters where words are to be searched for
+        wordDict: a dictionary containing methods isWord(string) and isPrefix(string)
+
+    Returns: 
+        ALL_WORDS: a set containing all valid words found. 
+    """
+    global CHAR_2D_ARRAY, WORD_DICT
+    CHAR_2D_ARRAY = char2dArray
+    WORD_DICT = wordDict
+    for i in range(row):
+        for j in range(col):
+            curr = CHAR_2D_ARRAY[i][j]
+            if WORD_DICT.isPrefix(curr):
+                visited = set()
+                visited.add((row, col))
+                findAllWordsFromCell(i, j, curr, visited)
+    return ALL_WORDS
+
+def findAllWordsFromCell(i, j, curr, visited):
+    """Find all words that begins with curr. 
+
+    Args:
+        i: row index of the current cell
+        j: col index of the current cell
+        curr: the current prefix string
+        visited: all cells visited when forming curr
+
+    Returns: 
+        Nothing. Note that ALL_WORDS is modified on the go. 
+    """
+    if WORD_DICT.isWord(curr):
+        ALL_WORDS.add(curr)
+    if i > 0:
+        checkLeftRight(i-1, j, curr, visited)
+        newCurr = curr + CHAR_2D_ARRAY[i-1][j]
+        checkCell(i-1, j, newCurr, visited)
+    if i < len(CHAR_2D_ARRAY) - 1:
+        checkLeftRight(i+1, j, curr, visited)
+        newCurr = curr + CHAR_2D_ARRAY[i+1][j]
+        checkCell(i+1, j, newCurr, visited)
+    checkLeftRight(i, j, curr, visited)
+
+def checkLeftRight(i, j, curr, visited):
+    """Check cells to the left and right of CHAR_2D_ARRAY[i][j]. 
+
+    Args:
+        i: row index of the current cell
+        j: col index of the current cell
+        curr: the current prefix string
+        visited: all cells visited when forming curr
+
+    Returns: 
+        Nothing. Note that ALL_WORDS is modified on the go. 
+    """
+    if j > 0:
+        newCurr = curr + CHAR_2D_ARRAY[i-1][j-1]
```

Comment:
It looks like it should cause the ~same computation to be run in the same way, just expressed more concisely in Python?

[List comprehensions](https://docs.python.org/2/tutorial/datastructures.html#list-comprehensions) can sometimes used to elegantly express a transformation (this looks like a fine example) but if one needs to change the code later they have a tendency to get complicated.

Pragmatically I'd suggest splitting a list comprehension into explicit for loops when it stops being immediately obvious what it does and you need to start reasoning about it to make sense of it. At code review time, if someone asks how a comprehension works just split it up instead/in addition to explaining it :)